pkg/controllers/resourceapply/config_hash.go

```diff
@@ -0,0 +1,136 @@
+package resourceapply
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/json"
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const configsHashAnnotation = "operator.openshift.io/configs-hash"
+
+type configSources struct {
+	ConfigMaps sets.Set[string]
+	Secrets    sets.Set[string]
+}
+
+// collectRelatedConfigSources looks into pod template spec for secret or config map references.
+// Currently, checks volumes and env vars for each container,
+// returns configSources structure which contains sets of config maps and secrets names.
+func collectRelatedConfigSources(spec *v1.PodTemplateSpec) configSources {
+	sources := configSources{
+		ConfigMaps: sets.Set[string]{},
+		Secrets:    sets.Set[string]{},
+	}
+
+	if spec == nil {
+		return sources
+	}
+
+	for _, volume := range spec.Spec.Volumes {
+		if volume.ConfigMap != nil {
+			sources.ConfigMaps.Insert(volume.ConfigMap.Name)
+		}
+		if volume.Secret != nil {
+			sources.Secrets.Insert(volume.Secret.SecretName)
+		}
+	}
+
+	for _, initContainer := range spec.Spec.InitContainers {
+		collectRelatedConfigsFromContainer(&initContainer, &sources)
+	}
+
+	for _, container := range spec.Spec.Containers {
+		collectRelatedConfigsFromContainer(&container, &sources)
+	}
+
+	return sources
+}
+
+// collectRelatedConfigsFromContainer collects related configs names into passed configSources instance.
+// Looks into env and envVar of the passed container spec and populates configSources with configmaps and secrets names.
+func collectRelatedConfigsFromContainer(container *v1.Container, sources *configSources) {
```

Comment:
I don't know that either of these inputs need to be pointers do they? We are only reading from the container and a configSources reference should still have the same pointers to the same maps as in the parent scope for the sets

Reply:
>I don't know that either of these inputs need to be pointers do they?

I made these pointers to avoid copying basically. For `configSources` I agree, it doesn't matter. For containers, hm, I'm not so sure.

Reply:
Looked into this again and recalled my thoughts a bit more:

`container` is a pointer to avoid unnecessary struct copying, and `sources` is mutating inside.


No strict necessity to have it as pointers, but I would like to leave it as is.

Reply:
IMO it's simpler for these not to be pointers, the data we are copying is relatively small and adding pointers adds complexity, as developers we now have to think about it more, so I would say it's preferable to be references on both. Go has a good GC, I'm not worried about data copying

Reply:
I do not agree that it adds complexity. 

First, this function is internal to the module and is not intended to be exposed.

Second, this code will be called for every reconciliation, and I do not want to have a `Container` copy there.

Third, pointer to `sources` + docstring indicates that `sources` will be mutated. Rely on the knowledge that there are sets inside and they are references - this adds complexity IMHO.

Reply:
I'm ok to continue with it as you prefer, but if this creates confusion in the future, I reserve the right to say I told you so 😉 
Agree to disagree as it were